File: neet_code_all/src/bin_search/single_element_in_sorted_array.rs
use std::collections::HashMap;

struct Solution {}
impl Solution {
    // 縛り無視ならビット演算で解くのが一番カンタン time: O(N), space: O(1)
    // AC
    pub fn single_non_duplicate_bit(nums: Vec<i32>) -> i32 {
        let mut result = 0;
        for v in nums {
            result ^= v;
        }

        result
    }

    // 縛り無視その２time: O(N), space: O(N)
    pub fn single_non_duplicate_map(nums: Vec<i32>) -> i32 {
        let mut map: HashMap<i32, i32> = HashMap::new();
        for v in nums {
            *map.entry(v).or_default() += 1;
        }

        let mut result = 0;
        for (v, n) in map {
            if n == 1 {
                result = v;
            }
        }

        result
    }

    // 縛りありでは解けなかった
    pub fn single_non_duplicate(nums: Vec<i32>) -> i32 {
        let (mut left, mut right) = (0_i32, nums.len() as i32 - 1);
        let mut mid = 0;
        while left <= right {
            mid = left + (right - left) / 2;
            if (mid - left + 1) % 2 == 0 {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        nums[mid as usize]
    }
}
// 縛り: time O(logN), space: O(1)

// C++の模範解答より
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    // AC
    pub fn single_non_duplicate(nums: Vec<i32>) -> i32 {
        let (mut left, mut right) = (0_i32, nums.len() as i32 - 2);

        while left <= right {
            let mid_1 = left + (right - left) / 2;
            let mid_2 = mid_1 ^ 1;

            if nums[mid_1 as usize] == nums[mid_2 as usize] {
                left = mid_1 + 1;
            } else {
                right = mid_1 - 1;
            }
        }

        nums[left as usize]
    }
}

fn main() {
    let case_1 = vec![1, 1, 2, 3, 3, 4, 4, 8, 8];
    // => 2
    let case_2 = vec![3, 3, 7, 7, 10, 11, 11];
    // => 10
    let case_3 = vec![1, 1, 2];
    // => 2

    println!(
        "case_1: {}",
        Solution::single_non_duplicate_bit(case_1.clone())
    );
    println!(
        "case_2: {}",
        Solution::single_non_duplicate_bit(case_2.clone())
    );
    println!(
        "case_3: {}",
        Solution::single_non_duplicate_bit(case_3.clone())
    );

    println!(
        "case_1: {}",
        Solution::single_non_duplicate_map(case_1.clone())
    );
    println!(
        "case_2: {}",
        Solution::single_non_duplicate_map(case_2.clone())
    );
    println!(
        "case_3: {}",
        Solution::single_non_duplicate_map(case_3.clone())
    );

    /*
    println!("case_1: {}", Solution::single_non_duplicate(case_1.clone()));
    println!("case_2: {}", Solution::single_non_duplicate(case_2.clone()));
    println!("case_3: {}", Solution::single_non_duplicate(case_3.clone()));
    */

    println!(
        "case_1: {}",
        SolutionAnsCpp::single_non_duplicate(case_1.clone())
    );
    println!(
        "case_2: {}",
        SolutionAnsCpp::single_non_duplicate(case_2.clone())
    );
    println!(
        "case_3: {}",
        SolutionAnsCpp::single_non_duplicate(case_3.clone())
    );
}

File: neet_code_all_cpp/bin_search/single_element_in_sorted_array.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

// 縛り: time O(logN), space: O(1)
class Solution {
    public:

        // 縛り無視(1): XORで調べる(time: O(N), space: O(1))
        // AC
        int singleNonDuplicateBit(std::vector<int>& nums) {
            int result = 0;
            for(auto& v: nums) {
                result ^= v;
            }

            return result;
        }

        // 縛り無視(2): mapを使う (time: O(N), space: O(N))
        // AC
        int singleNonDuplicateMap(std::vector<int>& nums) {
            std::unordered_map<int, int> map;
            for(auto& v: nums) {
                ++map[v];
            }

            int result = 0;
            for(auto& [v, n]: map) {
                if(n == 1) {
                    result = v;
                    break;
                }
            }

            return result;
        }
};

// 模範解答 
class SolutionAns {
    public:
        // 縛り通り
        int singleNonDuplicate(std::vector<int>& nums) {
            auto [left, right] = std::pair((int)0, (int)nums.size() - 2);
            while(left <= right) {
                auto mid_0 = left + (right - left) / 2;
                // mid_0が奇数なら1引く、偶数なら1足す
                auto mid_1 = mid_0 ^ 1;

                if(nums[mid_0] == nums[mid_1]) {
                    left = mid_0 + 1;
                } else {
                    right = mid_0 - 1;
                }
            }

            return nums[left];
        }

        // AC
        // XORを使うより２倍遅い
        int singleNonDuplicate2(std::vector<int>& nums) {
            auto [left, right] = std::pair((int)0, (int)nums.size() - 2);
            while(left <= right) {
                auto mid_0 = left + (right - left) / 2;
                // mid_0が奇数なら1引く、偶数なら1足す
                auto mid_1 = 0;
                if(mid_0 & 1) {
                    mid_1 = mid_0 - 1;
                } else {
                    mid_1 = mid_0 + 1;
                }

                if(nums[mid_0] == nums[mid_1]) {
                    left = mid_0 + 1;
                } else {
                    right = mid_0 - 1;
                }
            }

            return nums[left];
        }
};

int main(void) {
    std::vector<int> case_1 =  {1,1,2,3,3,4,4,8,8};
    // => 2
    std::vector<int> case_2 =  {3,3,7,7,10,11,11};
    // => 10

    Solution s_1;

    std::cout << s_1.singleNonDuplicateBit(case_1) << std::endl;
    std::cout << s_1.singleNonDuplicateBit(case_2) << std::endl;

    std::cout << s_1.singleNonDuplicateMap(case_1) << std::endl;
    std::cout << s_1.singleNonDuplicateMap(case_2) << std::endl;

    SolutionAns s_ans;

    std::cout << s_ans.singleNonDuplicate(case_1) << std::endl;
    std::cout << s_ans.singleNonDuplicate(case_2) << std::endl;

    std::cout << s_ans.singleNonDuplicate2(case_1) << std::endl;
    std::cout << s_ans.singleNonDuplicate2(case_2) << std::endl;
}
